libkbfs/file_data.go

```diff
@@ -509,44 +535,298 @@ func (fd *fileData) createIndirectBlock(
 	return fblock, nil
 }
 
-// newRightBlock appends a new block (with a temporary ID) to the end
-// of this file.
+// newRightBlock creates space for a new rightmost block, creating
+// parent blocks and a new level of indirection in the tree as needed.
+// If there's no new level of indirection, it modifies the blocks in
+// `parentBlocks` to include the new right-most pointers.  It also
+// returns the new top block (whether or not it is new), and also
+// returns any newly-dirtied block pointers.
 func (fd *fileData) newRightBlock(
-	ctx context.Context, ptr BlockPointer, pblock *FileBlock,
-	off int64) error {
-	newRID, err := fd.crypto.MakeTemporaryBlockID()
-	if err != nil {
-		return err
+	ctx context.Context, parentBlocks []parentBlockAndChildIndex, off int64,
+	df *dirtyFile, dver DataVer) (*FileBlock, []BlockPointer, error) {
```

Comment:
I think this function should, instead of modifying `parentBlocks`, return a new `[]parentBlockAndChildIndex` with the new pointers. It's less memory-efficient, but somewhat clearer, and we're already returning the first element of that array anyway (i.e., the new top block).